packages/arb-validator/rollupmanager/manager.go

```diff
@@ -38,20 +40,23 @@ const (
 	ValidEthBridgeVersion = "4"
 )
 
-type Manager struct {
-	// listenersLock is should be held whenever listeners is accessed or set
-	// and whenever activeChain is going to be set (but not when it is accessed)
-	listenersLock sync.Mutex
+var errNoActiveChain = errors.New("validator has no active chain")
+var notAtHead = errors.New("validator is catching up to latest")
+
+type reorgCache struct {
+	latestValidMachine    machine.Machine
+	currentPendingMachine machine.Machine
+	currentBlockId        *common.BlockId
+}
 
-	// validCallLock is held by the main loop whenever there is not a non-nil
-	// activeChain running which has caught up to the current L1 block. It is
-	// held by all other functions when they wish to access an update to date
-	// chain state. This means that these functions will block when there isn't
-	// an appropriate way to resolve them
-	validCallLock sync.Mutex
+type Manager struct {
+	// The mutex should be held whenever listeres or reorgCache are accessed or
+	// set and whenever activeChain is going to be set (but not when it is accessed)
+	sync.Mutex
 
 	listeners   []rollup.ChainListener
 	activeChain *rollup.ChainObserver
+	reorgCache  *reorgCache
```

Comment:
Should add a comment explaining what this is for and the limit on how it should be used.

Reply:
I added a lengthy comment